File: app/windows/WhisperOffline/UpdateChecker.cs
using System.Net.Http;
using System.Text.Json;

namespace WhisperOffline;

public record GhRelease(string Tag, string Name, string? ZipUrl, string Body);

/// Prüft GitHub-Releases auf neuere Versionen (öffentliches Repo, ohne Auth).
public static class UpdateChecker
{
    private const string LatestUrl =
        "https://api.github.com/repos/inf1nit3/whisper-offline/releases/latest";

    private static readonly HttpClient http = new();

    /// Installierte Version. Liest den ProductVersion aus der EXE, der
    /// <Version> im csproj folgt — beim Release nur eine Stelle bumpen.
    public static string CurrentVersion()
    {
        var path = Environment.ProcessPath;
        if (path != null && File.Exists(path))
        {
            var v = System.Diagnostics.FileVersionInfo.GetVersionInfo(path).ProductVersion;
            if (!string.IsNullOrEmpty(v)) return v.Split('+')[0];
        }
        return "0.0.0";
    }

    public static async Task<GhRelease?> FetchLatest()
    {
        try
        {
            using var req = new HttpRequestMessage(HttpMethod.Get, LatestUrl);
            req.Headers.Add("User-Agent", "whisper-offline-updater"); // GitHub verlangt UA
            req.Headers.Add("Accept", "application/vnd.github+json");
            using var resp = await http.SendAsync(req);
            if (!resp.IsSuccessStatusCode) return null;
            using var doc = JsonDocument.Parse(await resp.Content.ReadAsStringAsync());
            var root = doc.RootElement;

            string? zip = null;
            if (root.TryGetProperty("assets", out var assets))
                foreach (var a in assets.EnumerateArray())
                    if (a.GetProperty("name").GetString()?.EndsWith(".zip") == true)
                    {
                        zip = a.GetProperty("browser_download_url").GetString();
                        break;
                    }

            return new GhRelease(
                root.GetProperty("tag_name").GetString() ?? "",
                root.TryGetProperty("name", out var n) ? n.GetString() ?? "" : "",
                zip,
                root.TryGetProperty("body", out var b) ? b.GetString() ?? "" : "");
        }
        catch
        {
            return null;
        }
    }

    /// "v1.2.3" vs "1.2" → true wenn remote neuer.
    public static bool IsNewer(string current, string remoteTag)
    {
        var cur = Parse(current);
        var rem = Parse(remoteTag);
        for (int i = 0; i < 3; i++)
            if (rem[i] != cur[i]) return rem[i] > cur[i];
        return false;
    }

    private static int[] Parse(string v)
    {
        var clean = v.TrimStart('v', 'V').Split('+')[0];
        var parts = clean.Split('.', '-');
        return new[] { 0, 1, 2 }.Select(i => int.TryParse(parts.ElementAtOrDefault(i), out var n) ? n : 0).ToArray();
    }
}

File: app/android/app/src/main/java/dev/whisper/transcribe/UpdateChecker.kt
package dev.whisper.transcribe

import android.content.Context
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

/// Prüft GitHub-Releases auf neuere Versionen.
/// API: /repos/inf1nit3/whisper-offline/releases/latest (öffentlich, ohne Auth).
object UpdateChecker {

    const val REPO_LATEST =
        "https://api.github.com/repos/inf1nit3/whisper-offline/releases/latest"

    data class Release(
        val tag: String,          // z. B. "v1.1"
        val name: String,
        val apkUrl: String?,      // Download-URL des .apk-Assets
        val body: String,         // Release-Notes
    )

    fun currentVersion(context: Context): String =
        try {
            context.packageManager.getPackageInfo(context.packageName, 0).versionName ?: "0"
        } catch (_: Exception) { "0" }

    fun fetchLatest(): Release? {
        val conn = URL(REPO_LATEST).openConnection() as HttpURLConnection
        conn.connectTimeout = 10000
        conn.readTimeout = 15000
        conn.setRequestProperty("Accept", "application/vnd.github+json")
        conn.setRequestProperty("User-Agent", "whisper-offline-updater") // GitHub verlangt UA
        return try {
            if (conn.responseCode != 200) return null
            val root = JSONObject(conn.inputStream.bufferedReader().use { it.readText() })
            val apk = (0 until root.getJSONArray("assets").length())
                .map { root.getJSONArray("assets").getJSONObject(it) }
                .firstOrNull { it.getString("name").endsWith(".apk") }
                ?.getString("browser_download_url")
            Release(
                tag = root.getString("tag_name"),
                name = root.optString("name", root.getString("tag_name")),
                apkUrl = apk,
                body = root.optString("body", ""),
            )
        } catch (_: Exception) {
            null
        } finally {
            conn.disconnect()
        }
    }

    /// "v1.2.3" vs "1.2" → true, wenn remote neuere Versionsnummer hat.
    fun isNewer(current: String, remoteTag: String): Boolean {
        val cur = parse(current)
        val rem = parse(remoteTag)
        for (i in 0..2) if (rem[i] != cur[i]) return rem[i] > cur[i]
        return false
    }

    private fun parse(v: String): IntArray {
        val nums = v.trimStart('v', 'V').split('.', '-', '+')
        return IntArray(3) { i -> nums.getOrNull(i)?.toIntOrNull() ?: 0 }
    }
}
